hugr-passes/src/dataflow/partial_value.rs

```diff
@@ -0,0 +1,747 @@
+use ascent::lattice::BoundedLattice;
+use ascent::Lattice;
+use hugr_core::ops::Value;
+use hugr_core::types::{ConstTypeError, SumType, Type, TypeEnum, TypeRow};
+use hugr_core::{HugrView, Wire};
+use itertools::{zip_eq, Itertools};
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+
+use super::AbstractValue;
+
+/// Trait for abstract values that can be wrapped by [PartialValue] for dataflow analysis.
+/// (Allows the values to represent sums, but does not require this).
+pub trait BaseValue: Clone + std::fmt::Debug + PartialEq + Eq + Hash {
+    /// If the abstract value represents a [Sum] with a single known tag, deconstruct it
+    /// into that tag plus the elements. The default just returns `None` which is
+    /// appropriate if the abstract value never does (in which case [interpret_leaf_op]
+    /// must produce a [PartialValue::new_variant] for any operation producing
+    /// a sum).
+    ///
+    /// The signature is this way to optimize query/inspection (is-it-a-sum),
+    /// at the cost of requiring more cloning during actual conversion
+    /// (inside the lazy Iterator, or for the error case, as Self remains)
+    ///
+    /// [interpret_leaf_op]: super::DFContext::interpret_leaf_op
+    /// [Sum]: TypeEnum::Sum
+    /// [Tag]: hugr_core::ops::Tag
+    fn as_sum(&self) -> Option<(usize, impl Iterator<Item = Self> + '_)> {
```

Comment:
It'd be possible to use `supports_tag` and `variant_values` here, but that's more flexible than we need (this only needs to cope with concrete Sums, which have a single tag); and I think this single-method interface is much better. I tried to change `supports_tag`+`variant_values` into a single method returning an iterator but couldn't figure out anything (able to deal with multiple tags) that got past the borrow checker....